compiler/p/codegen/OMRTreeEvaluator.cpp

```diff
@@ -1027,7 +1027,44 @@ TR::Register *OMR::Power::TreeEvaluator::b2mEvaluator(TR::Node *node, TR::CodeGe
 
 TR::Register *OMR::Power::TreeEvaluator::s2mEvaluator(TR::Node *node, TR::CodeGenerator *cg)
 {
-    return TR::TreeEvaluator::unImpOpEvaluator(node, cg);
+    TR::Node *child = node->getFirstChild();
+    TR_ASSERT_FATAL(child->getDataType().getVectorElementType() == TR::Int64,
+        "Unsupported vector type %s for s2m (must be Int64)\n", child->getDataType().getVectorElementType().toString());
+
+    TR::Register *srcReg = cg->evaluate(child);
+    TR::Register *dstReg = cg->allocateRegister(TR_VRF);
+
+    TR::Register *tmpGPR = cg->allocateRegister(TR_GPR);
+    TR::Register *tmpVRF = cg->allocateRegister(TR_VRF);
+
+    node->setRegister(dstReg);
+
+    // rearrange byte elements of src to convert to word-length elements
+    generateTrg1Src1ImmInstruction(cg, TR::InstOpCode::sradi, node, tmpGPR, srcReg, 8);
+    generateTrg1Src1Imm2Instruction(cg, TR::InstOpCode::rldimi, node, tmpGPR, srcReg, 32, 0x000000FF00000000);
+
+    // move to VRF
+    generateTrg1Src1Instruction(cg, TR::InstOpCode::mtvsrd, node, dstReg, tmpGPR);
+
+    // unpack word-length elements to doubleword-length elements
+    generateTrg1Src1Instruction(cg, TR::InstOpCode::vupkhsw, node, dstReg, dstReg);
+
+    // reverse element order if big endian
+    if (!cg->comp()->target().cpu.isLittleEndian())
```

Comment:
Strictly speaking, we could avoid this xxpermdi and do slightly different masking and rotation at the beginning (without reversing the order).